tfhe/src/integer/ciphertext/compact_list.rs

```diff
@@ -1,42 +1,380 @@
+        if is_signed {
+            DataKind::Signed(num_blocks)
+        } else {
+            DataKind::Unsigned(num_blocks)
+        }
+    }
+}
+
+pub struct CompactCiphertextListBuilder {
+    messages: Vec<u64>,
+    info: Vec<DataKind>,
+    pub(crate) pk: CompactPublicKey,
+}
+
+impl CompactCiphertextListBuilder {
+    pub fn new(pk: &CompactPublicKey) -> Self {
+        Self {
+            messages: vec![],
+            info: vec![],
+            pk: pk.clone(),
+        }
+    }
+
+    pub fn push<T>(&mut self, data: T) -> &mut Self
+    where
+        T: Compactable,
+    {
+        let n = self.messages.len();
+        let kind = data.compact_into(&mut self.messages, self.pk.key.message_modulus(), None);
+        assert_eq!(n + kind.num_blocks(), self.messages.len());
+
+        if kind.num_blocks() != 0 {
+            self.info.push(kind);
+        }
+
+        self
+    }
+
+    pub fn push_with_num_blocks<T>(&mut self, data: T, num_blocks: usize) -> &mut Self
+    where
+        // The extra `Numeric` bound is to prevent T from being `bool`
+        T: Compactable + Numeric,
+    {
+        if num_blocks == 0 {
+            return self;
+        }
+
+        let n = self.messages.len();
+        let kind = data.compact_into(
+            &mut self.messages,
+            self.pk.key.message_modulus(),
+            Some(num_blocks),
+        );
+        assert_eq!(n + kind.num_blocks(), self.messages.len());
+        self.info.push(kind);
+        self
+    }
+
+    pub fn extend<T>(&mut self, values: impl Iterator<Item = T>) -> &mut Self
+    where
+        T: Compactable,
+    {
+        for value in values {
+            self.push(value);
+        }
+        self
+    }
+
+    pub fn extend_with_num_blocks<T>(
+        &mut self,
+        values: impl Iterator<Item = T>,
+        num_blocks: usize,
+    ) -> &mut Self
+    where
+        T: Compactable + Numeric,
+    {
+        for value in values {
+            self.push_with_num_blocks(value, num_blocks);
+        }
+        self
+    }
+
+    pub fn build(&self) -> CompactCiphertextList {
+        let ct_list = self.pk.key.encrypt_slice(self.messages.as_slice());
+        CompactCiphertextList {
+            ct_list,
+            info: self.info.clone(),
+        }
+    }
+
+    pub fn build_packed(&self) -> CompactCiphertextList {
+        // Here self.messages are decomposed blocks in range [0..message_modulus[
+        let msg_mod = self.pk.key.message_modulus().0 as u64;
+        let packed_messaged_iter = self
+            .messages
+            .chunks(2)
+            .map(|two_values| (two_values.get(1).copied().unwrap_or(0) * msg_mod) + two_values[0]);
+        let ct_list = self
+            .pk
+            .key
+            .encrypt_iter_with_modulus(packed_messaged_iter, msg_mod * msg_mod);
+
+        CompactCiphertextList {
+            ct_list,
+            info: self.info.clone(),
+        }
+    }
+
+    /// For zk proofs, message are always packed
```

Comment:
this is the docstring I was referring, we have one not packed here with a docstring saying it's always packed, and then the other function later in the file

Reply:
@tmontaigu don't know if you saw this, I think we have both functions and this docstring says zkpok are always packed

Reply:
Yep I will fix